zebra-network/src/protocol/external/types.rs

```diff
@@ -42,6 +47,41 @@ impl Default for Nonce {
     }
 }
 
+/// A random value to add to the seed value in a hash function.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Tweak(pub u32);
+
+impl Default for Tweak {
+    fn default() -> Self {
+        use rand::{thread_rng, Rng};
+        Self(thread_rng().gen())
+    }
+}
+
+/// A Bloom filter consisting of a bit field of arbitrary byte-aligned
+/// size, maximum size is 36,000 bytes.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Filter(pub Vec<u8>);
+
+impl ZcashSerialize for Filter {
+    fn zcash_serialize<W: Write>(&self, mut writer: W) -> Result<(), SerializationError> {
+        writer.write_all(&self.0)?;
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for Filter {
+    fn zcash_deserialize<R: Read>(reader: R) -> Result<Self, SerializationError> {
```

Comment:
This might be a bit extra but it makes read/write_filterload neater.

Reply:
Implementing these traits would be great but I am worried that we may not actually be able to, since per https://en.bitcoin.it/wiki/Protocol_documentation#filterload.2C_filteradd.2C_filterclear.2C_merkleblock it looks like there is both a) data after the filter data and b) no explicit length field, so my guess is that we may need to calculate the length from the total body size specified in the message header, minus 9 bytes for the trailing fields.

But this means that deserializing the filter blob requires additional contextual data, so we can't hope to do it as a `ZcashDeserialize/ZcashSerialize` pair, because those traits are context-free. Instead I think that we will have to pass the body length specified in the header into `read_filterload`, check that `9 ≤ len ≤ 36009`, and then compute the implicit length of the filter blob.

Reply:
Note: this is just based on my reading of that wiki page, maybe we should check zcashd?